### PR TITLE
[SPARK-36194][SQL] Add a logical plan visitor to propagate the distinct attributes

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

1. This pr add a new logical plan visitor named `DistinctKeyVisitor` to find out all the distinct attributes in current logical plan. For example:
   ```scala
   spark.sql("CREATE TABLE t(a int, b int, c int) using parquet")
   spark.sql("SELECT a, b, a % 10, a AS aliased_a, max(c), sum(b) FROM t GROUP BY a, b").queryExecution.analyzed.distinctKeys
   ```
   The output is: {a#1, b#2}, {b#2, aliased_a#0}.

2. Enhance `RemoveRedundantAggregates` to remove the aggregation if it is groupOnly and the child can guarantee distinct. For example:
   ```sql
   set spark.sql.autoBroadcastJoinThreshold=-1; -- avoid PushDownLeftSemiAntiJoin
   create table t1 using parquet as select id a, id as b from range(10);
   create table t2 using parquet as select id as a, id as b from range(8);
   select t11.a, t11.b from (select distinct a, b from t1) t11 left semi join t2 on (t11.a = t2.a) group by t11.a, t11.b;
   ```

   Before this PR:
   ```
   == Optimized Logical Plan ==
   Aggregate [a#6L, b#7L], [a#6L, b#7L], Statistics(sizeInBytes=1492.0 B)
   +- Join LeftSemi, (a#6L = a#8L), Statistics(sizeInBytes=1492.0 B)
      :- Aggregate [a#6L, b#7L], [a#6L, b#7L], Statistics(sizeInBytes=1492.0 B)
      :  +- Filter isnotnull(a#6L), Statistics(sizeInBytes=1492.0 B)
      :     +- Relation default.t1[a#6L,b#7L] parquet, Statistics(sizeInBytes=1492.0 B)
      +- Project [a#8L], Statistics(sizeInBytes=984.0 B)
         +- Filter isnotnull(a#8L), Statistics(sizeInBytes=1476.0 B)
            +- Relation default.t2[a#8L,b#9L] parquet, Statistics(sizeInBytes=1476.0 B)
   ```

   After this PR:
   ```
   == Optimized Logical Plan ==
   Join LeftSemi, (a#6L = a#8L), Statistics(sizeInBytes=1492.0 B)
   :- Aggregate [a#6L, b#7L], [a#6L, b#7L], Statistics(sizeInBytes=1492.0 B)
   :  +- Filter isnotnull(a#6L), Statistics(sizeInBytes=1492.0 B)
   :     +- Relation default.t1[a#6L,b#7L] parquet, Statistics(sizeInBytes=1492.0 B)
   +- Project [a#8L], Statistics(sizeInBytes=984.0 B)
      +- Filter isnotnull(a#8L), Statistics(sizeInBytes=1476.0 B)
         +- Relation default.t2[a#8L,b#9L] parquet, Statistics(sizeInBytes=1476.0 B)
   ```

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and TPC-DS benchmark test.

SQL | Before this PR(Seconds) | After this PR(Seconds)
-- | -- | --
q14a | 206  | 193
q38 | 59 | 41
q87 | 127 | 113

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlanDistinctKeys.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * A trait to add distinct attributes to [[LogicalPlan]]. For example:
+ * {{{
+ *   SELECT a, b, SUM(c) FROM Tab1 GROUP BY a, b
+ *   // returns a, b
+ * }}}
+ */
+trait LogicalPlanDistinctKeys { self: LogicalPlan =>
+  lazy val distinctKeys: Set[AttributeSet] = DistinctKeyVisitor.visit(self)
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlanDistinctKeys.scala:30]: e.g. `spark.sql.optimizer.propagateDistinctKeys.enabled`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    projectList.filter {
+      case a: Alias => expressions.exists(_.semanticEquals(a.child))
+      case ne => expressions.exists(_.semanticEquals(ne))
+    }.toSet.subsets(keys.map(_.size).min).filter { s =>
+      val references = s.map {
+        case a: Alias => a.child
+        case ne => ne
+      }
+      keys.exists(_.equals(ExpressionSet(references)))
+    }.map(s => AttributeSet(s.map(_.toAttribute))).toSet
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:40]: how about
```
    val outputSet = ExpressionSet(projectList.map(_.toAttribute))
    val aliases = projectList.filter(_.isInstanceOf[Alias])
    if (aliases.isEmpty) return keys.filter(_.subsetOf(outputSet))

    val aliasedDistinctKeys = keys.map { expressionSet =>
      expressionSet.map { expression =>
        expression transform {
          case expr: Expression =>
            aliases
              .collectFirst { case a: Alias if a.child.semanticEquals(expr) => a.toAttribute }
              .getOrElse(expr)
        }
      }
    }
    aliasedDistinctKeys.collect {
      case es: ExpressionSet if es.subsetOf(outputSet) => ExpressionSet(es)
    }
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:40]: If one expression has multiple aliases, we need to further expand the distinct keys set. We can do it later as it's rather a corner case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  override def visitExcept(p: Except): Set[AttributeSet] =
+    if (!p.isAll && p.deterministic) Set(p.outputSet) else default(p)
+
+  override def visitExpand(p: Expand): Set[AttributeSet ] = default(p)
+
+  override def visitFilter(p: Filter): Set[AttributeSet ] = p.child.distinctKeys
+
+  override def visitGenerate(p: Generate): Set[AttributeSet ] = default(p)
+
+  override def visitGlobalLimit(p: GlobalLimit): Set[AttributeSet ] = p.child.distinctKeys
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:63]: if the limit value is 1 or 0, all output columns are distinct.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+  override def visitIntersect(p: Intersect): Set[AttributeSet ] = {
+    if (!p.isAll && p.deterministic) Set(p.outputSet) else default(p)
+  }
+
+  override def visitJoin(p: Join): Set[AttributeSet] = {
+    p.joinType match {
+      case LeftExistence(_) => p.left.distinctKeys
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:71]: shall we exclude `ExistenceJoin`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      case _ => default(p)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:72]: for inner join, the distinct keys can be the join keys.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      default(p)
+    }
+  }
+
+  override def visitRepartition(p: Repartition): Set[AttributeSet] = p.child.distinctKeys
+
+  override def visitRepartitionByExpr(p: RepartitionByExpression): Set[AttributeSet] =
+    p.child.distinctKeys
+
+  override def visitSample(p: Sample): Set[AttributeSet] = default(p)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:93]: For Sample without replacement, we can propagate the distinct keys from child.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  override def visitWindow(p: Window): Set[AttributeSet] = p.child.distinctKeys
+
+  override def visitTail(p: Tail): Set[AttributeSet] = p.child.distinctKeys
+
+  override def visitSort(p: Sort): Set[AttributeSet] = p.child.distinctKeys
+
+  override def visitRebalancePartitions(p: RebalancePartitions): Set[AttributeSet] =
+    p.child.distinctKeys
+
+  override def visitWithCTE(p: WithCTE): Set[AttributeSet] = default(p)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:108]: CTE can also propagate distinct keys from child.

**Comment by cloud-fan** [comment]
due to significant plan changes caused by this PR, can we verify the TPCDS performance?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeSet, ExpressionSet, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.LeftExistence
+
+/**
+ * A visitor pattern for traversing a [[LogicalPlan]] tree and propagate the distinct attributes.
+ */
+object DistinctKeyVisitor extends LogicalPlanVisitor[Set[AttributeSet]] {
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:26]: Change `[Set[AttributeSet]]` to `[Set[ExpressionSet]]` to fix the join condition contains complex expressions:
```scala
sql("create table t1 using parquet as select id a, id as b from range(10)")
sql("create table t2 using parquet as select id as a, id as b from range(8)")
println(sql("select * from (select distinct a, b from t1) t11 inner join (select distinct a, b from t2) t22 on (t11.a = t22.a and t11.b % 100 = t22.b % 100) ").queryExecution.optimizedPlan.distinctKeys)
```
The output should be empty.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 7 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 6 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 5 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:93]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 4 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 2 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlanDistinctKeys.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlanDistinctKeys.scala]: OK

**Comment by wangyum** [comment]
> due to significant plan changes caused by this PR, can we verify the TPCDS performance?

OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala [hunk 9]
```diff
@@ -47,6 +47,10 @@ object RemoveRedundantAggregates extends Rule[LogicalPlan] with AliasHelper {
       } else {
         newAggregate
       }
+
+     case agg @ Aggregate(groupingExps, _, child) if agg.groupOnly && child.deterministic &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala:51]: does `child.deterministic` matter here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+  private def projectDistinctKeys(
+      keys: Set[ExpressionSet], projectList: Seq[NamedExpression]): Set[ExpressionSet] = {
+    val outputSet = ExpressionSet(projectList.map(_.toAttribute))
+    val distinctKeys = keys.filter(_.subsetOf(outputSet))
+    val aliases = projectList.filter(_.isInstanceOf[Alias])
+    if (aliases.isEmpty) {
+      distinctKeys
+    } else {
+      val aliasedDistinctKeys = keys.map { expressionSet =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:37]:
```suggestion
      val aliasedDistinctKeys = distinctKeys.map { expressionSet =>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    if (!p.isAll && p.deterministic) Set(ExpressionSet(p.output)) else default(p)
+  }
+
+  override def visitJoin(p: Join): Set[ExpressionSet] = {
+    p.joinType match {
+      case LeftSemiOrAnti(_) => p.left.distinctKeys
+      case Inner =>
+        p match {
+          case ExtractEquiJoinKeys(_, leftKeys, rightKeys, _, _, _, _, _)
+              if p.left.distinctKeys.exists(_.subsetOf(ExpressionSet(leftKeys))) &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:88]: I think we should use `||` here. If we only one side has valid distinct keys, we should still propagate that side.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                p.right.distinctKeys.exists(_.subsetOf(ExpressionSet(rightKeys))) =>
+            Set(ExpressionSet(leftKeys), ExpressionSet(rightKeys))
+          case _ => default(p)
+        }
+      case _ => default(p)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:93]: for left outer, we can propagate from right side.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 13]
```diff
@@ -3888,6 +3897,8 @@ class SQLConf extends Serializable with Logging {
 
   def constraintPropagationEnabled: Boolean = getConf(CONSTRAINT_PROPAGATION_ENABLED)
 
+  def propagateDistinctKeysEnabled: Boolean = getConf(PROPAGATE_DISTINCT_KEYS_ENABLED)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3900]: it's only called once, we can inline it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala [hunk 14]
```diff
@@ -166,12 +162,137 @@ class RemoveRedundantAggregatesSuite extends PlanTest {
...
+      .analyze
+    val correctAnswer = relation
+      .groupBy('a)('a, count('b).as("cnt"))
+      .select('a, 'cnt)
+      .analyze
+    val optimized = Optimize.execute(originalQuery)
+    comparePlans(optimized, correctAnswer)
+  }
+
+  test("SPARK-36194: Negative case: The grouping expressions not same") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala:246]: we don't require them to be the same. We need the child distinct keys to be a subset of the required grouping keys.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala [hunk 15]
```diff
@@ -166,12 +162,137 @@ class RemoveRedundantAggregatesSuite extends PlanTest {
...
+      val optimized = Optimize.execute(originalQuery.analyze)
+      comparePlans(optimized, originalQuery.analyze)
+    }
+  }
+
+  test("SPARK-36194: Negative case: The aggregate expressions not the sub aggregateExprs") {
+    Seq(LeftSemi, LeftAnti).foreach { joinType =>
+      val originalQuery = x.groupBy('a, 'b)('a, 'b)
+        .join(y, joinType, Some("x.a".attr === "y.a".attr && "x.b".attr === "y.b".attr))
+        .groupBy("x.a".attr, "x.b".attr)(TrueLiteral)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala:261]: hmmm why can't we optimize this query?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala [hunk 16]
```diff
@@ -166,12 +162,137 @@ class RemoveRedundantAggregatesSuite extends PlanTest {
...
+
+      val optimized = Optimize.execute(originalQuery.analyze)
+      comparePlans(optimized, originalQuery.analyze)
+    }
+  }
+
+  test("SPARK-36194: Negative case: The aggregate expressions not same") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala:268]: the test name is a bit misleading. I don't think aggregate expressions matter (as long as it's a group only aggregate), grouping expressions matter/

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala (hunk 9 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala:51]: We had a test like this before:
https://github.com/apache/spark/blob/a34e2afb7538894f9a264b894924e848cd6979e0/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala#L104-L114

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 11 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:88]:
```scala
spark.sql("create table t1(a int, b int) using parquet")
spark.sql("create table t2(x int, y int) using parquet")

spark.sql("insert into t1 values(1, 1), (2, 2)")
spark.sql("insert into t2 values(1, 1), (1, 1), (2, 2), (2, 2)")

spark.sql("select * from (select distinct * from t1 )t1 join t2 on t1.a = t2.x and t1.b = t2.y").show
```
The output is:
```
+---+---+---+---+
|  a|  b|  x|  y|
+---+---+---+---+
|  1|  1|  1|  1|
|  1|  1|  1|  1|
|  2|  2|  2|  2|
|  2|  2|  2|  2|
+---+---+---+---+
```

We can't distinguish the distinct keys.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 12 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:93]:
```scala
spark.sql("create table t1(a int, b int) using parquet")
spark.sql("create table t2(x int, y int) using parquet")

spark.sql("insert into t1 values(1, 1), (2, 2)")
spark.sql("insert into t2 values(3, 3), (4, 4)")

spark.sql("select * from t1 left join (select distinct * from t2)t2 on t1.a = t2.x and t1.b = t2.y").show
```
The output is:
```
+---+---+----+----+
|  a|  b|   x|   y|
+---+---+----+----+
|  2|  2|null|null|
|  1|  1|null|null|
+---+---+----+----+
```

We can't distinguish the distinct keys.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 13 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3900]: OK

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala (hunk 14 above)
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala:246]: Yes. update the test name.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala (hunk 15 above)
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala]: Yes. Update `RemoveRedundantAggregates` to support this case:
https://github.com/apache/spark/blob/f0db8ecb810b446a48a870e55b2394d3202ae8b7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala#L56-L60

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala (hunk 16 above)
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala]: Fixed the test name.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala (hunk 9 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala]: which means `child.deterministic` doesn't matter? The test you posted did optimize out one aggregate.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala]: Sorry. This test:
https://github.com/apache/spark/blob/f0db8ecb810b446a48a870e55b2394d3202ae8b7/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesSuite.scala#L164-L171

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala [hunk 17]
```diff
@@ -47,6 +47,17 @@ object RemoveRedundantAggregates extends Rule[LogicalPlan] with AliasHelper {
...
         newAggregate
       }
+
+     case agg @ Aggregate(groupingExps, _, child)
+         if agg.groupOnly && child.deterministic &&
+           child.distinctKeys.exists(_.subsetOf(ExpressionSet(groupingExps))) =>
+      Project(agg.aggregateExpressions, child)
+
+    case agg @ Aggregate(groupingExps, aggregateExps, child)
+        if aggregateExps.forall(a => a.isInstanceOf[Alias] && a.children.forall(_.foldable)) &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala:57]: how about a mix? e.g. `SELECT a, 1, b FROM ... GROUP BY a, b, c`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  override def visitJoin(p: Join): Set[ExpressionSet] = {
+    p match {
+      case Join(_, _, LeftSemiOrAnti(_), _, _) =>
+        p.left.distinctKeys
+      case ExtractEquiJoinKeys(Inner, leftKeys, rightKeys, _, _, _, _, _)
+        if p.left.distinctKeys.exists(_.subsetOf(ExpressionSet(leftKeys))) &&
+          p.right.distinctKeys.exists(_.subsetOf(ExpressionSet(rightKeys))) =>
+        Set(ExpressionSet(leftKeys), ExpressionSet(rightKeys))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:89]: `if p.left.distinctKeys.exists(_.subsetOf(ExpressionSet(leftKeys)))`, we can propagate the right side distinct keys, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 12 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:90]: Sorry my fault. We can propagate the left side distinct keys if `p.right.distinctKeys.exists(_.subsetOf(rightJoinKeySet))`

**Comment by wangyum** [comment]
New PR: https://github.com/apache/spark/pull/35779

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 18 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:89]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala (hunk 17 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala:57]: We already support this case, add a new test case:
https://github.com/apache/spark/pull/35779/files#diff-7cd933ffc7b9ce86d5973bee80f4a5bd4a021c0f0ff81defe1f020bcb55b4b3bR153-R159

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala (hunk 12 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:90]: +1

**Pull request closed without merging**